File: sdk/python/control_licencia.py
"""SDK de licencias CONTROL para Python / FastAPI (Avendia).

Requiere `cryptography` y `httpx` (ambos ya están en Avendia).

    from control_licencia import ControlLicencia
    licencia = ControlLicencia(
        url=settings.control_url, clave=settings.control_licencia, producto="avendia",
        clave_publica_base64=settings.control_clave_publica, archivo="datos/licencia.json",
        huella=settings.dominio_publico, version=settings.version,
    )

    @app.on_event("startup")
    async def _licencia():
        estado = await licencia.iniciar()
        if not estado["valido"]:
            raise RuntimeError(f"Licencia no válida: {estado['motivo']}")

    app.add_middleware(licencia.middleware_asgi())        # 402 si deja de ser válida
"""
from __future__ import annotations

import base64
import hashlib
import json
import socket
import time
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

import httpx
from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey


def _b64url_decode(s: str) -> bytes:
    return base64.urlsafe_b64decode(s + "=" * (-len(s) % 4))


class ControlLicencia:
    def __init__(self, url: str, clave: str, producto: str, clave_publica_base64: str, archivo: str = "licencia.json",
                 huella: str | None = None, version: str | None = None, latido_horas: float = 24) -> None:
        self.url = url.rstrip("/")
        self.clave = clave.strip().upper()
        self.producto = producto
        self.publica = Ed25519PublicKey.from_public_bytes(base64.b64decode(clave_publica_base64))
        self.archivo = Path(archivo)
        self.version = version
        self.huella = huella or hashlib.sha256(socket.gethostname().encode()).hexdigest()[:32]
        self.latido_segundos = latido_horas * 3600
        self.estado: dict[str, Any] = {"valido": False, "payload": None, "motivo": "sin iniciar"}
        self._ultimo_latido = 0.0

    def verificar(self, token: str) -> dict | None:
        if not isinstance(token, str) or "." not in token:
            return None
        cuerpo, firma = token.split(".", 1)
        try:
            self.publica.verify(_b64url_decode(firma), cuerpo.encode())
            p = json.loads(_b64url_decode(cuerpo))
        except (InvalidSignature, ValueError):
            return None
        if p.get("clave") != self.clave or p.get("producto") != self.producto or p.get("huella") != self.huella:
            return None
        if datetime.fromisoformat(p["expira_en"].replace("Z", "+00:00")) < datetime.now(timezone.utc):
            return None
        return p

    async def _llamar(self, ruta: str, cuerpo: dict) -> dict:
        async with httpx.AsyncClient(timeout=10) as cliente:
            r = await cliente.post(f"{self.url}/api/v1/licencias/{ruta}", json={k: v for k, v in cuerpo.items() if v is not None})
        try:
            datos = r.json()
        except ValueError:
            datos = {}
        return {"ok": r.is_success, "status": r.status_code, **datos}

    def _procesar(self, r: dict) -> dict:
        if r.get("ok") and r.get("token"):
            p = self.verificar(r["token"])
            if p:
                try:
                    self.archivo.parent.mkdir(parents=True, exist_ok=True)
                    self.archivo.write_text(json.dumps({"token": r["token"], "guardado_en": datetime.now(timezone.utc).isoformat()}))
                except OSError:
                    pass
                self.estado = {"valido": True, "payload": p, "motivo": None}
                return self.estado
        if r.get("status") in (403, 404):
            self.estado = {"valido": False, "payload": r.get("licencia"), "motivo": r.get("error") or r.get("codigo") or "licencia inválida", "codigo": r.get("codigo")}
        return self.estado

    async def activar(self) -> dict:
        return self._procesar(await self._llamar("activar", {
            "clave": self.clave, "producto": self.producto, "huella": self.huella,
            "dominio": self.huella, "nombre_equipo": socket.gethostname(), "version": self.version,
        }))

    async def latido(self) -> dict:
        self._ultimo_latido = time.time()
        return self._procesar(await self._llamar("latido", {"clave": self.clave, "huella": self.huella, "version": self.version}))

    async def iniciar(self) -> dict:
        """Token guardado si es válido; si no, activa. Sin red, vale hasta expira_en."""
        guardado = None
        try:
            guardado = json.loads(self.archivo.read_text()).get("token")
        except (OSError, ValueError):
            pass
        p = self.verificar(guardado) if guardado else None
        if p:
            self.estado = {"valido": True, "payload": p, "motivo": None}
        try:
            await (self.latido() if p else self.activar())
        except httpx.HTTPError as e:
            if not p:
                self.estado = {"valido": False, "payload": None, "motivo": f"Sin conexión con CONTROL: {e}"}
        return self.estado

    def vigente(self) -> bool:
        p = self.estado.get("payload") or {}
        expira = p.get("expira_en")
        if self.estado.get("valido"):
            return True
        return bool(expira) and datetime.fromisoformat(expira.replace("Z", "+00:00")) > datetime.now(timezone.utc)

    def middleware_asgi(self, rutas_libres: tuple[str, ...] = ("/api/v1/salud", "/docs", "/openapi.json")):
        """Middleware ASGI puro: responde 402 cuando la licencia no está vigente y renueva el latido."""
        licencia = self

        class Middleware:
            def __init__(self, app):
                self.app = app

            async def __call__(self, scope, receive, send):
                if scope["type"] != "http" or scope["path"].startswith(rutas_libres):
                    return await self.app(scope, receive, send)
                if time.time() - licencia._ultimo_latido > licencia.latido_segundos:
                    try:
                        await licencia.latido()
                    except httpx.HTTPError:
                        pass
                if not licencia.vigente():
                    cuerpo = json.dumps({"error": "Licencia no válida", "motivo": licencia.estado.get("motivo"), "codigo": licencia.estado.get("codigo")}).encode()
                    await send({"type": "http.response.start", "status": 402, "headers": [(b"content-type", b"application/json")]})
                    await send({"type": "http.response.body", "body": cuerpo})
                    return
                return await self.app(scope, receive, send)

        return Middleware

File: sdk/node/control-licencia.js
/**
 * SDK de licencias CONTROL para Node.js (Express, Next.js, NestJS, Electron).
 * Sin dependencias: usa fetch y node:crypto.
 *
 *   import { ControlLicencia } from './control-licencia.js';
 *   const licencia = new ControlLicencia({
 *     url: process.env.CONTROL_URL,              // https://control.tuagencia.com
 *     clave: process.env.CONTROL_LICENCIA,       // CTL-XXXX-XXXX-XXXX-XXXX
 *     producto: 'dental-pro',                    // código del producto en CONTROL
 *     clavePublica: 'BASE64_DE_32_BYTES',        // GET /api/v1/licencias/clave-publica
 *     archivo: './datos/licencia.json',          // dónde guardar el último token válido
 *     version: '1.4.0',
 *   });
 *   await licencia.iniciar();                    // activa o valida offline; lanza si está bloqueada
 *   app.use(licencia.middlewareExpress());       // 402 si la licencia deja de ser válida
 */
import { createPublicKey, createHash, verify } from 'node:crypto';
import { readFileSync, writeFileSync, mkdirSync } from 'node:fs';
import { dirname } from 'node:path';
import { hostname } from 'node:os';

const PREFIJO_SPKI_ED25519 = Buffer.from('302a300506032b6570032100', 'hex');

export class ControlLicencia {
  constructor({ url, clave, producto, clavePublica, archivo = './licencia.json', version, huella, dominio, latidoHoras = 24 }) {
    if (!url || !clave || !producto || !clavePublica) throw new Error('CONTROL: faltan url, clave, producto o clavePublica');
    this.url = url.replace(/\/$/, '');
    this.clave = clave.trim().toUpperCase();
    this.producto = producto;
    this.publica = createPublicKey({ key: Buffer.concat([PREFIJO_SPKI_ED25519, Buffer.from(clavePublica, 'base64')]), format: 'der', type: 'spki' });
    this.archivo = archivo;
    this.version = version;
    this.dominio = dominio;
    this.huella = huella || ControlLicencia.huellaPorDefecto(dominio);
    this.latidoMs = latidoHoras * 3600000;
    this.estado = { valido: false, payload: null, motivo: 'sin iniciar' };
  }

  /** Huella estable: dominio (web) o hash de hostname (escritorio). */
  static huellaPorDefecto(dominio) {
    return createHash('sha256').update(dominio || hostname()).digest('hex').slice(0, 32);
  }

  /** Verifica un token con la clave pública. Devuelve el payload o null. */
  verificar(token) {
    if (typeof token !== 'string' || !token.includes('.')) return null;
    const [cuerpo, firma] = token.split('.');
    if (!verify(null, Buffer.from(cuerpo), this.publica, Buffer.from(firma, 'base64url'))) return null;
    try {
      const p = JSON.parse(Buffer.from(cuerpo, 'base64url').toString('utf8'));
      if (p.clave !== this.clave || p.producto !== this.producto || p.huella !== this.huella) return null;
      if (Date.parse(p.expira_en) < Date.now()) return null;
      return p;
    } catch { return null; }
  }

  leerGuardado() { try { return JSON.parse(readFileSync(this.archivo, 'utf8')).token; } catch { return null; } }
  guardar(token) { try { mkdirSync(dirname(this.archivo), { recursive: true }); writeFileSync(this.archivo, JSON.stringify({ token, guardado_en: new Date().toISOString() })); } catch { /* disco de solo lectura: se sigue en memoria */ } }

  async llamar(ruta, cuerpo) {
    const r = await fetch(`${this.url}/api/v1/licencias/${ruta}`, { method: 'POST', headers: { 'content-type': 'application/json' }, body: JSON.stringify(cuerpo) });
    const datos = await r.json().catch(() => ({}));
    return { ok: r.ok, status: r.status, ...datos };
  }

  /** Activa (o re-valida) contra el servidor. */
  async activar() {
    const r = await this.llamar('activar', { clave: this.clave, producto: this.producto, huella: this.huella, dominio: this.dominio, nombre_equipo: hostname(), version: this.version });
    return this.procesarRespuesta(r);
  }

  async latido() {
    const r = await this.llamar('latido', { clave: this.clave, huella: this.huella, version: this.version });
    return this.procesarRespuesta(r);
  }

  procesarRespuesta(r) {
    if (r.ok && r.token) {
      const p = this.verificar(r.token);
      if (p) { this.guardar(r.token); this.estado = { valido: true, payload: p, motivo: null }; return this.estado; }
    }
    // Respuesta del servidor con bloqueo explícito (suspendida, revocada, max_activaciones…).
    if (r.status === 403 || r.status === 404) {
      this.estado = { valido: false, payload: r.licencia || null, motivo: r.error || r.codigo || 'licencia inválida', codigo: r.codigo };
    }
    return this.estado;
  }

  /**
   * Arranque: usa el token guardado si sigue válido; si no, intenta activar.
   * Sin red y con token válido, funciona hasta `expira_en` (gracia offline).
   */
  async iniciar() {
    const guardado = this.leerGuardado();
    const p = guardado && this.verificar(guardado);
    if (p) this.estado = { valido: true, payload: p, motivo: null };
    try { await (p ? this.latido() : this.activar()); }
    catch (e) { if (!p) this.estado = { valido: false, payload: null, motivo: `Sin conexión con CONTROL: ${e.message}` }; }
    if (this.latidoMs > 0) { this.temporizador = setInterval(() => this.latido().catch(() => {}), this.latidoMs); this.temporizador.unref?.(); }
    if (!this.estado.valido) throw new Error(`Licencia no válida: ${this.estado.motivo}`);
    return this.estado;
  }

  detener() { if (this.temporizador) clearInterval(this.temporizador); }

  /** Middleware Express/Next (API routes): bloquea con 402 si la licencia dejó de ser válida. */
  middlewareExpress({ rutasLibres = ['/salud', '/licencia'] } = {}) {
    return (req, res, next) => {
      if (rutasLibres.some((r) => req.path.startsWith(r))) return next();
      const p = this.estado.payload;
      if (this.estado.valido || (p && Date.parse(p.expira_en) > Date.now())) {
        if (p?.estado === 'mora') res.set('X-Licencia-Aviso', 'Licencia vencida: renueva antes de que se suspenda');
        return next();
      }
      res.status(402).json({ error: 'Licencia no válida', motivo: this.estado.motivo, codigo: this.estado.codigo });
    };
  }
}
